src/broker/index.js

```diff
@@ -62,16 +73,55 @@ module.exports = class Broker {
     this.sessionLifetime = Long.ZERO
 
     // The lock timeout has twice the connectionTimeout because the same timeout is used
-    // for the first apiVersions call
-    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
+    // for the first apiVersions call and it has additional little buffer to last little longer
+    // so when {this.authLock} timeouts within connection, the auth timeout error will be created, not generic lock timeout.
+    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout + 50
     this.brokerAddress = `${this.connection.host}:${this.connection.port}`
 
     this.lock = new Lock({
       timeout: lockTimeout,
       description: `connect to broker ${this.brokerAddress}`,
     })
 
+    this.authLock = new Lock({
+      timeout: this.authenticationTimeout,
+      description: `authentication to broker ${this.brokerAddress}`,
+    })
+
     this.lookupRequest = notInitializedLookup
+
+    /**
+     * @private
+     * @returns {Promise}
+     */
+    this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
+      try {
+        await this.authLock.acquire()
```

Comment:
We don't need this lock anymore now right? If there's more callers, the first one will create a promise and start the authentication, the second one will just receive the promise from the first one. Since the promise creation itself is synchronous there shouldn't be any race condition there, I think.

Reply:
yes you are right. I will remove auth lock.